bolt-design-system/bolt
# Video integration bugfixes

## Jira

- http://vjira2:8080/browse/BDS-1339

## Summary

This fixes video issues in IE11 reported by @remydenton while integrating 2.3.1 with Drupal.

This branch is based on the v2.3.1 tag. I created a separate branch and PR based on master: https://github.com/bolt-design-system/bolt/pull/1183

## Details

### Overview
- Cherry-pick Symbol polyfill hotfix
- Rework Video component JS to prevent bugs that result from videos initializing multiple times

### Comments
@remydenton reported a bug where, in IE11 on a page with multiple instances of the same video with email plugin enabled, only the last video got the email plugin. See [a demo](https://demo6.d8.pega.com/content-page/test-videos).

I couldn't reproduce this in Pattern Lab, and I tried really hard. I could, however, reproduce it in Drupal locally. My best guess is that there's a race condition, surfaced by the environmental differences, that causes it to break in Drupal and not Pattern Lab.

I applied some of the Video implementation changes from the "[Convert Video to Web Component](https://github.com/bolt-design-system/bolt/pull/1099)" branch, where I faced similar bug, and that fixed the plugin issue for me locally in Drupal.

### Implementation details
Brightcove changed their best practices for [referencing a video player](https://support.brightcove.com/using-bc-and-getplayer-methods) and for [delayed player instantiation](https://support.brightcove.com/delaying-player-instantiation) around v6.12 - v6.18. Our implementation needed to be updated to adhere to the new standards to avoid the confusing initialization issues mentioned in BC's [release notes](https://support.brightcove.com/brightcove-player-release-notes-0) (see 6.18.3).

Since the code change on Brightcove's side, many of our videos worked fine, but we've been plagued by intermittent bugs, especially around plugins. These updates should fix that.

The key changes are:
 - Don't include `data-account`, `data-player`, or `data-video-id` attributes on first render of the `<video-js>` element or videos will auto-initialize. Add them on `rendered()` just before calling `bc()`. ([code](https://github.com/bolt-design-system/bolt/pull/1180/files#diff-dc990b7be4fb8a542dcb25a607c1a370R661))
 - Check if video has already initialized before calling `bc()` to prevent re-initializing the video on re-render when attributes are added. ([code](https://github.com/bolt-design-system/bolt/pull/1180/files#diff-dc990b7be4fb8a542dcb25a607c1a370R570))
 - Don't use `data-setup`. It's a Video.js prop that will confuse Brightcove. Instead pass those configuration options to the Brightcove function itself. ([code](https://github.com/bolt-design-system/bolt/pull/1180/files#diff-dc990b7be4fb8a542dcb25a607c1a370R584))
 - Use `<video-js>` not `<video>`, new best practice from Brightcove to avoid flash of native video controls. ([code](https://github.com/bolt-design-system/bolt/pull/1180/files#diff-dc990b7be4fb8a542dcb25a607c1a370R657))
 - Reference video player with `videojs.getPlayer(id)` not `videojs(id)`. The latter can cause Video.js to initialize. ([code](https://github.com/bolt-design-system/bolt/pull/1180/files#diff-2a95e839f21b89378583b5065482ce33R22))

## How to test
*Waiting for built to finish. Links are to master for now.*
- Review the [files changed](https://github.com/bolt-design-system/bolt/pull/1180/files).
- Test [all videos](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video/index.html) in Chrome, IE11 and Edge.
- Make sure if the video is supposed to have a plugin, that plugin appears.
- Check the social share overlay, especially on:
  - [Custom Share](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video-55-video-with-custom-share-text/02-components-video-55-video-with-custom-share-text.html)
  - [Default Plugins](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video-plugins-default-plugins/02-components-video-plugins-default-plugins.html)
  - [Disabled Plugins](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video-plugins-disabled-plugins/02-components-video-plugins-disabled-plugins.html)
  - [Enabled Plugins](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video-plugins-enabled-plugins/02-components-video-plugins-enabled-plugins.html)
  - [External Plugins Script](https://master.boltdesignsystem.com/pattern-lab/patterns/02-components-video-plugins-external-plugins-script/02-components-video-plugins-external-plugins-script.html)
